### PR TITLE
Use type traits in linear solving

### PR DESCRIPTION
**Pull request by joschmitt**
So far this does not change the functionality. I open this pull request already now to discuss names and the design.

I will add the new Howell form to the system and try to clean up more of the `_can_solve...` methods in `src/Matrix.jl` (no promises on this though).

Oh, and this will break Nemo in the current state.

### CODE REVIEW

File: src/Solve.jl [hunk 1]
```diff
@@ -190,9 +220,20 @@ function solve_context_type(::Type{T}) where {T <: Union{Rational{BigInt}, FracE
   return SolveCtx{T, dense_matrix_type(T), IntMatT, IntMatT}
 end
 
+matrix_normal_form_type(C::SolveCtx) = matrix_normal_form_type(base_ring(C))
+
+# Overwrite RrefTrait for fields
+matrix_normal_form_type(C::SolveCtx{<:FieldElement}) = LUTrait()
+function matrix_normal_form_type(C::SolveCtx{T}) where {T <: Union{Rational{BigInt}, FracElem}}
```
Comment (thofma) [inline src/Solve.jl:227]: Doesn't it pick the right one for `FracElem` anyway?

**Comment by thofma** [comment]
https://github.com/Nemocas/AbstractAlgebra.jl/actions/runs/9975920868/job/27580200035?pr=1738 should be working?

**Comment by joschmitt** [comment]
> https://github.com/Nemocas/AbstractAlgebra.jl/actions/runs/9975920868/job/27580200035?pr=1738 should be working?

I don't fully understand what's going on there. An Aqua test in Nemo is failing and I have the impression that it is testing Nemo against the released version of AbstractAlgebra. Then this
```
WARNING: could not import AbstractAlgebra.howell_form! into Nemo
ERROR: LoadError: UndefVarError: `matrix_normal_form_type` not defined
```
would be expected.

**Comment by lgoettgens** [comment]
> > [Nemocas/AbstractAlgebra.jl/actions/runs/9975920868/job/27580200035?pr=1738](https://github.com/Nemocas/AbstractAlgebra.jl/actions/runs/9975920868/job/27580200035?pr=1738) should be working?
> 
> I don't fully understand what's going on there. An Aqua test in Nemo is failing and I have the impression that it is testing Nemo against the released version of AbstractAlgebra. Then this
> 
> ```
> WARNING: could not import AbstractAlgebra.howell_form! into Nemo
> ERROR: LoadError: UndefVarError: `matrix_normal_form_type` not defined
> ```
> 
> would be expected.

This is probably due to https://github.com/JuliaTesting/Aqua.jl/issues/292

**Review by thofma — APPROVED** [review]

**Pull request merged**